src/Stratis.Features.FederatedPeg/TargetChain/MaturedBlocksSyncManager.cs

```diff
@@ -97,24 +109,83 @@ public class MaturedBlocksSyncManager : IMaturedBlocksSyncManager
+            /// <para>
+            /// In a production/live scenario the sidechain and mainnet federation nodes should run on the same machine.
+            /// </para>
+            /// </summary>
+            public ConsensusClient(ICounterChainSettings counterChainSettings, IHttpClientFactory httpClientFactory)
+                : base(httpClientFactory, counterChainSettings.CounterChainApiPort, "Consensus", $"http://{counterChainSettings.CounterChainApiHost}")
+            {
+            }
+
+            public Task<List<ThresholdActivationModel>> GetLockedInDeployments(CancellationToken cancellation = default)
+            {
+                return this.SendGetRequestAsync<List<ThresholdActivationModel>>("lockedindeployments", null, cancellation);
+            }
+        }
+
+        public void RecordCounterChainActivations()
+        {
+            // If this is the main chain then ask the side-chain for its activation height.
+            if (this.network.Name.StartsWith("Cirrus"))
+                return;
+
+            if (this.mainChainActivationHeight != int.MaxValue)
+                return;
+
+            ConsensusClient consensusClient = new ConsensusClient(this.counterChainSettings, this.httpClientFactory);
+            List<ThresholdActivationModel> lockedInActivations = consensusClient.GetLockedInDeployments(this.nodeLifetime.ApplicationStopping).ConfigureAwait(false).GetAwaiter().GetResult();
+            if (lockedInActivations == null || lockedInActivations.Count == 0)
+                return;
+
+            ThresholdActivationModel model = lockedInActivations.FirstOrDefault(a => a.DeploymentName.ToLower() == "release1300");
+            if (model == null || model.lockedInTimestamp == null)
+                return;
+
+            if (this.chainIndexer.Tip.Header.Time < model.lockedInTimestamp.Value)
+                return;
+
+            int mainChainLockedInHeight = this.chainIndexer.Tip.EnumerateToGenesis().TakeWhile(h => h.Header.Time > (uint)(model.lockedInTimestamp)).FirstOrDefault().Height;
+
+            Network counterChainNetwork = this.counterChainSettings.CounterChainNetwork;
+            this.mainChainActivationHeight = mainChainLockedInHeight + 
+                (int)((counterChainNetwork.Consensus.MinerConfirmationWindow * counterChainNetwork.Consensus.TargetSpacing.TotalSeconds) / this.network.Consensus.TargetSpacing.TotalSeconds);
+        }
+
+        public int GetMainChainActivationHeight()
+        {
+            return this.mainChainActivationHeight;
+        }
+
         /// <inheritdoc />
         public async Task StartAsync()
         {
             // Initialization delay; give the counter chain node some time to start it's API service.
             await Task.Delay(TimeSpan.FromSeconds(InitializationDelaySeconds), this.nodeLifetime.ApplicationStopping).ConfigureAwait(false);
 
+            RecordCounterChainActivations();
```

Comment:
We dont need to call this here if we call this in the loop anyway ;)

Reply:
I was just worried that someone may implement logic that depends on these values before the loop starts running. So upon exit from `StartAsync` I ideally want the values already set.